pkg/common/resources/dra.go

```diff
@@ -60,3 +67,123 @@ func RemoveReservedFor(claim *resourceapi.ResourceClaim, pod *v1.Pod) {
+// and returns a ResourceList with GPU resources aggregated.
+func ExtractDRAGPUResources(ctx context.Context, pod *v1.Pod, kubeClient client.Client) (v1.ResourceList, error) {
+	gpuResources := v1.ResourceList{}
+
+	if len(pod.Spec.ResourceClaims) == 0 {
+		return gpuResources, nil
+	}
+
+	// Map to group claims by DeviceClassName and count devices
+	deviceClassCounts := make(map[string]int64)
+
+	for _, podClaim := range pod.Spec.ResourceClaims {
+		claimName, err := GetResourceClaimName(pod, &podClaim)
+		if err != nil {
+			// Skip claims that don't have a name yet (e.g., from templates not yet created)
+			continue
+		}
+
+		claim := &resourceapi.ResourceClaim{}
+		claimKey := types.NamespacedName{
+			Namespace: pod.Namespace,
+			Name:      claimName,
+		}
+
+		err = kubeClient.Get(ctx, claimKey, claim)
+		if err != nil {
+			// Skip claims that don't exist yet or can't be fetched
+			continue
+		}
+
+		gpuCount, err := countGPUDevicesFromClaim(claim)
+		if err != nil {
+			// Skip invalid claims but continue processing others
+			continue
+		}
+
+		if gpuCount > 0 {
+			// Find the DeviceClassName for this claim
+			deviceClassName := getGPUDeviceClassNameFromClaim(claim)
+			if deviceClassName != "" {
+				deviceClassCounts[deviceClassName] += gpuCount
+			}
+		}
+	}
+
+	// Convert aggregated counts to ResourceList
+	if len(deviceClassCounts) > 0 {
+		totalGPUs := int64(0)
+		for _, count := range deviceClassCounts {
+			totalGPUs += count
+		}
+		if totalGPUs > 0 {
+			gpuResources[constants.GpuResource] = *resource.NewQuantity(totalGPUs, resource.DecimalSI)
+		}
+	}
+
+	return gpuResources, nil
+}
+
+// isGPUDeviceClass checks if a DeviceClassName represents a GPU.
+// Checks for "nvidia.com/gpu" and also accepts any device class name containing "gpu"
+// (case-insensitive) to support custom GPU device classes like "gpu.example.com".
+func isGPUDeviceClass(deviceClassName string) bool {
+	if deviceClassName == constants.GpuResource {
+		return true
+	}
+	// Check if device class name contains "gpu" (case-insensitive) to support custom GPU device classes
+	deviceClassNameLower := strings.ToLower(deviceClassName)
+	return strings.Contains(deviceClassNameLower, "gpu")
+}
+
+// getGPUDeviceClassNameFromClaim extracts the GPU DeviceClassName from a ResourceClaim.
+// Returns empty string if no GPU device class is found.
+func getGPUDeviceClassNameFromClaim(claim *resourceapi.ResourceClaim) string {
+	for _, request := range claim.Spec.Devices.Requests {
+		if request.Exactly != nil && isGPUDeviceClass(request.Exactly.DeviceClassName) {
+			return request.Exactly.DeviceClassName
+		}
+	}
+	return ""
+}
+
+// countGPUDevicesFromClaim counts GPU devices from a ResourceClaim.
+// Returns the total count of GPU devices requested by this claim.
+func countGPUDevicesFromClaim(claim *resourceapi.ResourceClaim) (int64, error) {
+	totalCount := int64(0)
+
+	for _, request := range claim.Spec.Devices.Requests {
+		if request.Exactly == nil {
```

Comment:
We should remember that we should count `firstAvailable` requests in the future, maybe open a ticket?